docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+* vectorized loop that operates on multiple values at a time
+* handling of the remainder
+
+Example: our input is a buffer of ten integers, assuming that `Vector128` is accelerated, we handle the first four values in the first loop iteration, the next four in the second iteration and then we stop, as only two are left. Depending on how we can handle the remainder, we distinguish two approaches.
+
+### Scalar remainder handling
+
+Imagine that we want to calculate the sum of all the numbers in given buffer. We definitely want to add every element just once, without repetitions. That is why in the first loop, we add four (128 bits / 32 bits) integers in one iteration. In the second loop, we handle the remaining values.
+
+
+```cs
+int Sum(Span<int> buffer)
+{
+    Debug.Assert(Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<int>.Count);
+
+    // The initial sum is zero, so we need a vector with all elements initialized to zero.
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    // We need to obtain the reference to first value in the buffer, it's used later for loading vectors from memory.
+    ref int searchSpace = ref MemoryMarshal.GetReference(buffer);
+    // And an offset, that is going to be used by vectorized and scalar loops.
+    nuint elementOffset = 0;
+    // And the last valid offset from which we can load the values
+    nuint oneVectorAwayFromEnd = (nuint)(buffer.Length - Vector128<int>.Count);
+    for (; elementOffset <= oneVectorAwayFromEnd; elementOffset += (nuint)Vector128<int>.Count)
+    {
+        // We load a vector from given offset.
+        Vector128<int> loaded = Vector128.LoadUnsafe(ref searchSpace, elementOffset);
+        // We add 4 integers at a time:
+        sum += loaded;
+    }
+
+    // We sum all 4 integers from the vector to one
+    int result = Vector128.Sum(sum);
+
+    // And handle the remaining elements, in a non-vectorized way:
+    while (elementOffset < (nuint)buffer.Length)
+    {
+        result += buffer[(int)elementOffset];
+        elementOffset++;
+    }
+
+    return result;
+}
+```
+
+**Note:** Use `ref MemoryMarshal.GetReference(span)` instead of `ref span[0]` and `ref MemoryMarshal.GetArrayDataReference(array)` instead of `ref array[0]` to handle empty buffer scenarios (which would throw `IndexOutOfRangeException`). If the buffer is empty, these methods return a reference to the location where the 0th element would have been stored. Such a reference may or may not be null. You can use it for pinning but you must never dereference it.
+
+**Note:** The `GetReference` method has an overload that accepts a `ReadOnlySpan` and returns mutable reference. Please use it with caution! To get a `readonly` reference, you need to use [ReadOnlySpan<T>.GetPinnableReference](https://learn.microsoft.com/dotnet/api/system.readonlyspan-1.getpinnablereference).
```

Comment:
You don't "need" to use `GetPinnableReference`. The latter just guarantees you only have a `readonly` reference.

You can still do `ref readonly T searchSpace = ref MemoryMarshal.GetReference(buffer)` to get a readonly reference otherwise.